src/server/conn.rs

```diff
@@ -49,9 +49,17 @@ impl JoinQueue {
             joined = self.join_next() => joined,
             _ = time::sleep(timeout) => {
                 if self.queue.len() > 1 {
-                    self.queue.swap(0, 1);
+                    // If the connection task can't be joined in `timeout`,
+                    // move it to the back of the queue. It might be using
+                    // a websocket.
+                    match self.queue.pop_front() {
+                        Some(handle) => self.queue.push_back(handle),
+                        None => panic!("join_next was not canceled safely"),
```

Comment:
The implementation of `join_next` is potentially not cancel safe. A little bit of work is required to ensure that we are joining the same join handle when the task is woken up by the waker.

Perhaps, the `idle` field can be changed to an `Option<Id>` to ensure that the same task will be polled when the waker wakes up the task to continue making progress.

From what I can tell, there is little risk associated with this in the current implementation. It seems like the worst case scenario would be polling a different join handle at the front of the queue if `wake_by_ref` is called just before the timeout expires.